packages/gatsby/src/bootstrap/redirects-writer.ts

```diff
@@ -8,15 +8,59 @@ import { IRedirect } from "../redux/types"
 let lastHash: string | null = null
 let bootstrapFinished = false
 
+function normalizePathPrefix(pathPrefix: string): string {
+  const newPathPrefix = pathPrefix.endsWith(`/`)
+    ? pathPrefix.slice()
+    : `${pathPrefix}/`
+
+  // strip trailing `/`
+  return newPathPrefix.replace(/\/$/, ``)
+}
+
+export function ensureLeadingSlash(path: string): string {
+  if (!path.startsWith(`/`) && !path.startsWith(`http`)) {
+    return `/` + path
+  }
+  return path
+}
+
+function normalizePath(path: string, pathPrefix: string | null): string {
+  let normalizedPath
+
+  if (pathPrefix && !path.startsWith(`http`)) {
+    const newPathPrefix = normalizePathPrefix(pathPrefix)
+    normalizedPath = path.replace(newPathPrefix, ``)
+  } else {
+    normalizedPath = path
+  }
+
+  return ensureLeadingSlash(normalizedPath)
+}
```

Comment:
Where is this code coming from?

Why does the path need to be normalized / have a trailing slash?

Why is `decodeURIComponent` used and not the `URL` constructor?

Reply:
This logic is copied over to mirror the validation that already exists in Cloud when deploying redirect routes. I'm also considering whether this would be handled better somewhere earlier in the flow there, as @TylerBarnes made a good point that this change is tied to Gatsby Cloud builds and not necessarily OSS.